internal/feedback/feedback.go

```diff
@@ -25,23 +28,50 @@ const DepTypeImported = "imported dep"
 
 // ConstraintFeedback holds project constraint feedback data
 type ConstraintFeedback struct {
-	Version, LockedVersion, Revision, ConstraintType, DependencyType, ProjectPath string
+	Constraint, LockedVersion, Revision, ConstraintType, DependencyType, ProjectPath string
 }
 
-// LogFeedback logs the feedback
+// NewConstraintFeedback builds a feedback entry for a constraint in the manifest.
+func NewConstraintFeedback(c gps.Constraint, pr gps.ProjectRoot, depType string) *ConstraintFeedback {
+	cf := &ConstraintFeedback{
+		Constraint:     c.String(),
+		ProjectPath:    string(pr),
+		DependencyType: depType,
+	}
+
+	if _, ok := c.(gps.Revision); ok {
+		cf.ConstraintType = ConsTypeHint
+	} else {
+		cf.ConstraintType = ConsTypeConstraint
+	}
+
+	return cf
+}
+
+// NewLockedProjectFeedback builds a feedback entry for a project in the lock.
+func NewLockedProjectFeedback(v gps.Version, pr gps.ProjectRoot, depType string) *ConstraintFeedback {
+	cf := &ConstraintFeedback{
+		ProjectPath:    string(pr),
+		DependencyType: depType,
+	}
+
+	switch vt := v.(type) {
+	case gps.PairedVersion:
+		cf.LockedVersion = vt.String()
+		cf.Revision = vt.Underlying().String()
+	case gps.Revision:
+		cf.Revision = vt.String()
+	}
```

Comment:
It should be, as a practical matter, impossible to have a `gps.UnpairedVersion` here. But it's still logically possible, so we need to have a case for it all the same. 

```go
case gps.UnpairedVersion:
	cf.LockedVersion = vt.String()
```